src/main/java/com/marklogic/contentpump/ThreadManager.java

```diff
@@ -0,0 +1,372 @@
+                    LOG.debug("Running with MultithreadedMapper. Initial " +
+                        "thread count for split #" + index + ": " +
+                        taskThreads);
+                }
+            }
+        }
+        taskList.add(task);
+        if (runtimeMapperClass == (Class)MultithreadedMapper.class) {
+            synchronized (pool) {
+                taskFutureList.add(pool.submit(task));
+                pool.wait();
+            }
+        } else {
+            pool.submit(task);
+        }
+    }
+
+    /**
+     * Assign thread count for a given split
+     *
+     * @param splitIndex split index
+     * @param splitCount
+     * @return number of threads for each input split
+     */
+    private int assignThreads(int splitIndex, int splitCount, int totalThreads) {
+        if (threadsPerSplit > 0) {
+            return threadsPerSplit;
+        }
+        if (splitCount == 1) {
+            return totalThreads;
+        }
+        if (splitCount * minThreads > totalThreads) {
+            return minThreads;
+        }
+        if (splitIndex % totalThreads < totalThreads % splitCount) {
+            return totalThreads / splitCount + 1;
+        } else {
+            return totalThreads / splitCount;
+        }
+    }
+
+    public static void setRestrictHosts(boolean newRestrictHosts) {
+        restrictHosts = newRestrictHosts;
+    }
+
+    /**
+     * Run server thread polling query and adjust thread pool
+     */
+    class ThreadPoller implements Runnable {
+        @Override
+        public void run() {
+            if (ContentPump.shutdown) {
+                scheduler.shutdown();
+                return;
+            }
+            if (!runAutoScaling()) return;
+            // Poll server max threads
+            String[] hosts = conf.getStrings(MarkLogicConstants.OUTPUT_HOST);
+            for (int i = 0; i < hosts.length; i++) {
+                try {
+                    ContentSource cs = InternalUtilities.getOutputContentSource(
+                        conf, hosts[i]);
+                    queryServerMaxThreads(cs);
+                    if (LOG.isDebugEnabled()) {
+                        LOG.debug("New available server threads: " +
+                            newServerThreads);
+                    }
+                    break;
+                } catch (Exception e) {
+                    if (e.getCause() instanceof ServerConnectionException) {
+                        LOG.warn("Unable to connect to " + hosts[i]
+                            + " to query destination information");
+                        if (LOG.isDebugEnabled()) {
+                            LOG.debug(e);
+                        }
+                    } else {
+                        LOG.error(e.getMessage(), e);
+                        job.setJobState(JobStatus.State.FAILED);
+                        return;
+                    }
+                }
+            }
+
+            if (curServerThreads < newServerThreads) {
```

Comment:
Just confirming, if an exception is thrown, curServerThreads would be equal to newServerThreads, and no action is taken, right?